app/src/main/java/com/upbad/apps/opgo/plugin/ClockStyleOp12.java

```diff
@@ -0,0 +1,136 @@
+package com.upbad.apps.opgo.plugin;
+
+import static com.upbad.apps.opgo.util.ReflectionUtil.log;
+
+import android.content.Context;
+import android.text.SpannableString;
+import android.text.SpannableStringBuilder;
+import android.text.TextUtils;
+import android.view.ViewGroup;
+import android.widget.TextView;
+
+import de.robv.android.xposed.XC_MethodHook;
+import de.robv.android.xposed.XposedHelpers;
+import de.robv.android.xposed.callbacks.XC_LoadPackage;
+
+public class ClockStyleOp12 implements IPlugin {
+
+    @Override
+    public void hook(XC_LoadPackage.LoadPackageParam lpparam, ClassLoader classLoader) {
+        try {
+            log("ClockStyle OOS12 hook() started");
+            // on OnePlus OxygenOS 12, the logic changed
+            // notification clock
+            Class<?> kClockExt = classLoader.loadClass("com.oplusos.systemui.ext.ClockExt");
+            // target method: com.oplusos.systemui.ext.ClockExt->setTextWithRedOneStyleInternal(TextView, CharSequence)V
+            XposedHelpers.findAndHookMethod(kClockExt, "setTextWithRedOneStyleInternal",
+                    TextView.class, CharSequence.class, new XC_MethodHook() {
+                        @Override
+                        protected void afterHookedMethod(MethodHookParam param) {
+                            TextView textView = (TextView) param.args[0];
+                            CharSequence text = (CharSequence) param.args[1];
+                            textView.setText(text.toString().replace(":", "∶"));
+                        }
+                    });
+
+            // keyguard digital clock
+            // com.oplusos.keyguard.OpKeyguardOplusTextView->getTextWithOplusColor(CharSequence)SpannableString
+            Class<?> kOpKeyguardOplusTextView = classLoader.loadClass("com.oplusos.keyguard.OpKeyguardOplusTextView");
+            XposedHelpers.findAndHookMethod(kOpKeyguardOplusTextView, "getTextWithOplusColor",
+                    CharSequence.class, new XC_MethodHook() {
+                        @Override
+                        protected void beforeHookedMethod(MethodHookParam param) {
+                            CharSequence text = (CharSequence) param.args[0];
+                            SpannableString spannableString = new SpannableString(text.toString());
+                            param.setResult(spannableString);
+                        }
+                    });
+
+            // on OOS 12, the AOD is a standalone app whose package name is "com.oplus.aod"
+            // the ClassLoader of this app is not the same as the ClassLoader of the main app
+            // entry point: com.oplusos.systemui.aod.plugin.AodForeignPackage->initForeignPackage()Z
+            XposedHelpers.findAndHookMethod("com.oplusos.systemui.aod.plugin.AodForeignPackage", classLoader,
```

Comment:
Thanks for the PR!

As you mentioned that AOD is a standalone app in OOS 12, should the parameter `classLoader` be instantiated from the AOD package, instead of the one from "com.android.systemui"?

Reply:
The `AodForeignPackage.initForeignPackage()` is the AOD plugin loader which is invoked in the host(systemui) process. **This means the AOD is loaded into the systemui process(as a plugin),** rather than in it's own app process. Although the AOD.apk itself is a standalone app, when AOD take effect the AOD is working like a plugin in the host process(systemui).

Reply:
Make sense. Thanks for your explanation!